### PR TITLE
Preparation for OrderedRDD2.

### PR DESCRIPTION
**Pull request by cseed**
Use off-heap storage.
Replaced Platform => is.hail.annotations.Memory.
Offsets are now Long, including stored offests to binary and arrays.

### CODE REVIEW

**Comment by cseed** [comment]
@tpoterba Ready for another look.  I now use the coders to serialize UnsafeRow and UnsafeIndexedSeq.

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 1]
```diff
@@ -177,19 +185,22 @@ class Encoder(out: DataOutputStream) {
     out.writeDouble(d)
   }
 
-  def writeBytes(b: Array[Byte], off: Int, n: Int) {
-    out.write(b, off, n)
+  def writeBytes(mem: Long, off: Long, n: Int) {
+    var i = 0
+    while (i < n) {
+      out.writeByte(Memory.loadByte(mem + off + i))
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: is there any way to make this more efficient?

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 2]
```diff
@@ -1,73 +1,99 @@
...
   }
+
+  override def write(kryo: Kryo, output: Output) {
+    kryo.writeObject(output, ttBc)
+    val enc = new Encoder(new DataOutputStream(output))
+    enc.writeRegionValue(ttBc.value.typ, region, aoff)
+  }
+
+  override def read(kryo: Kryo, input: Input) {
+    val cTorrentBroadcast = Class.forName("org.apache.spark.broadcast.TorrentBroadcast")
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: why do you need this? Can't you just use the readObject method?

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 3]
```diff
@@ -117,7 +143,10 @@ object UnsafeRow {
         readArray(region, offset, elementType, ttBc).toSet
       case TString => readString(region, offset)
       case td: TDict =>
-        readArray(region, offset, td.elementType, ttBc).asInstanceOf[IndexedSeq[Row]].map(r => (r.get(0), r.get(1))).toMap
+        println(td)
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: stray println

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 4]
```diff
@@ -117,7 +143,10 @@ object UnsafeRow {
...
+        val a = readArray(region, offset, td.elementType, ttBc)
+        println(a)
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: stray println

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 5]
```diff
@@ -226,4 +255,18 @@ class UnsafeRow(val ttBc: Broadcast[TypeTree],
...
   }
+
+  override def write(kryo: Kryo, output: Output) {
+    kryo.writeObject(output, ttBc)
+    val enc = new Encoder(new DataOutputStream(output))
+    enc.writeRegionValue(ttBc.value.typ, region, offset)
+  }
+
+  override def read(kryo: Kryo, input: Input) {
+    val cTorrentBroadcast = Class.forName("org.apache.spark.broadcast.TorrentBroadcast")
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: same serialization concern

File: src/main/scala/is/hail/annotations/Memory.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+  def storeInt(addr: Long, i: Int): Unit = unsafe.putInt(addr, i)
+
+  def storeLong(addr: Long, l: Long): Unit = unsafe.putLong(addr, l)
+
+  def storeFloat(addr: Long, f: Float): Unit = unsafe.putFloat(addr, f)
+
+  def storeDouble(addr: Long, d: Double): Unit = unsafe.putDouble(addr, d)
+
+  def storeAddress(addr: Long, a: Long): Unit = unsafe.putAddress(addr, a)
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/Memory.scala]: On a 32-bit machine it is a 32-bit store.  We're allocating 64-bits in memory for addresses always, but we should be using Unsafe.ADDRESS_SIZE to allocate pointers (if we want to support 32-bit machines.  Meh.)

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala (hunk 1 above)
Comment (cseed) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: Yes, but it is some work and I just want to get it working.

File: src/main/scala/is/hail/annotations/UnsafeRow.scala (hunk 2 above)
Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: readObject needs a type, but TorrentBroadcast is private.

File: src/main/scala/is/hail/annotations/Memory.scala [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+  def malloc(size: Long): Long = unsafe.allocateMemory(size)
+
+  def free(a: Long): Unit = unsafe.freeMemory(a)
+
+  def realloc(a: Long, newSize: Long): Long = unsafe.reallocateMemory(a, newSize)
+
+  def memcpy(dst: Long, src: Long, n: Long): Unit = unsafe.copyMemory(src, dst, n)
+
+  def copyToArray(dst: Array[Byte], src: Long, n: Long): Unit = {
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/Memory.scala]: Fixed.  At first I didn't understand how to use Unsafe to copy between on- and off-heap.

**Comment by cseed** [comment]
Addressed comments.  @tpoterba back to you.

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 8]
```diff
@@ -284,10 +301,10 @@ class RichRDDRow(val rdd: RDD[Row]) extends AnyVal {
           rvb.addRow(t, r)
           val offset = rvb.end()
 
-          val rowSize = region.offset
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: shouldn't this be a long?

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 9]
```diff
@@ -1,73 +1,99 @@
...
+    ttBc = kryo.readObject(input, cTorrentBroadcast).asInstanceOf[Broadcast[TypeTree]]
+    val t = ttBc.value.typ.asInstanceOf[TArray]
+
+    val dec = new Decoder(new DataInputStream(input))
+    region = MemoryBuffer()
+    aoff = dec.readRegionValue(t, region)
+
+    length = region.loadInt(aoff)
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: should be a long

Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: I didn't change array length to Long yet.  Right now arrays are limited to max Int in length.

**Comment by cseed** [comment]
Back to you @tpoterba

**Pull request merged**